src/main/java/com/endava/internship/collections/StudentMap.java

```diff
@@ -1,78 +1,248 @@
+        }
+        else {
+            return findKey(current.getLeft(), key);
+        }
+    }
+    public boolean findValue(Node current, Object value, Boolean isPresent){
+        if (current != null){
+            if (current.getValue().equals(value)){
+                return true;
+            }else {
+                isPresent = findValue(current.getLeft(), value, isPresent);
+                isPresent = findValue(current.getRight(), value, isPresent);
+                return isPresent;
+            }
+        }else{
+            return isPresent;
+        }
+    }
+    private Integer removeByKey(Object o) {
+        Integer value;
+        Node parentNode = getNodeReference(head, head, o);
+
+        if (head.getKey().equals(o)){
+            value = head.getValue();
+            Node newHead = head.getLeft();
+            Node rightNode = newHead.getRight();
+            newHead.setRight(head.getRight());
+            head = newHead;
+            head = insert(head, rightNode);
+            size -= 2;
+            return value;
+        }
+
+        Node nodeToDelete;
+        if (parentNode.getLeft() == null){
+            nodeToDelete = parentNode.getRight();
+            parentNode.setRight(null);
+        }else if (parentNode.getRight() == null){
+            nodeToDelete = parentNode.getLeft();
+            parentNode.setLeft(null);
+        }else {
+            if (parentNode.getRight().getKey().equals(o)){
+                nodeToDelete = parentNode.getRight();
+                parentNode.setRight(null);
+            }
+            else {
+                nodeToDelete = parentNode.getLeft();
+                parentNode.setLeft(null);
+            }
+        }
+
+        Node leftRemain = nodeToDelete.getLeft();
+        Node rightRemain = nodeToDelete.getRight();
+        value = nodeToDelete.getValue();
+
+        if (leftRemain == null && rightRemain == null){
+            size--;
+            return value;
+        }
+
+        if (leftRemain == null) {
+            head = insert(head, rightRemain);
+            size -= 2;
+        }
+        else if(rightRemain == null) {
+            head = insert(head, leftRemain);
+            size -= 2;
+        }
+        else {
+            head = insert(head, leftRemain);
+            head = insert(head, rightRemain);
+            size -= 3;
+        }
+        return value;
+    }
+    private Node getNodeReference(Node current,Node previous, Object key){
+        if (current != null){
+            if (current.getKey().equals(key))
+                return previous;
+            else if (current.getKey().compareTo((Student) key) > 0){
+                return getNodeReference(current.getRight(),current, key);
+            }else {
+                return getNodeReference(current.getLeft(),current, key);
+            }
+        }else {
+            return null;
+        }
+    }
+    private Node clearTree(Node current){
+        if (current == null)
+            return null;
```

Comment:
refactor this method, it returns null either way

Reply:
Fixed.